R/parametric.R

```diff
@@ -0,0 +1,363 @@
+      ))
+      params[[i]]$pdf = pdf
+      params[[i]]$cdf = cdf
+      params[[i]]$quantile = quantile
+    }
+  } else if (form == "ph") {
+    for (i in seq_along(lp)) {
+      body(pdf) = substitute((exp(y) * basedist$hazard(x)) * (1 - self$cdf(x)), list(y = -lp[i]))
+      body(cdf) = substitute(1 - (basedist$survival(x)^exp(y)), list(y = -lp[i]))
+      body(quantile) = substitute(
+        basedist$quantile(1 - exp(exp(-y) * log(1 - p))), # nolint
+        list(y = -lp[i])
+      )
+      params[[i]]$pdf = pdf
+      params[[i]]$cdf = cdf
+      params[[i]]$quantile = quantile
+    }
+  } else if (form == "aft") {
+    for (i in seq_along(lp)) {
+      body(pdf) = substitute((exp(-y) * basedist$hazard(x / exp(y))) * (1 - self$cdf(x)),
+        list(y = lp[i]))
+      body(cdf) = substitute(1 - (basedist$survival(x / exp(y))), list(y = lp[i]))
+      body(quantile) = substitute(exp(y) * basedist$quantile(p), list(y = lp[i]))
+      params[[i]]$pdf = pdf
+      params[[i]]$cdf = cdf
+      params[[i]]$quantile = quantile
+    }
+  } else if (form == "po") {
+    for (i in seq_along(lp)) {
+      body(pdf) = substitute((basedist$hazard(x) *
+        (1 - (basedist$survival(x) /
+          (((exp(y) - 1)^-1) + basedist$survival(x))))) *
+        (1 - self$cdf(x)), list(y = lp[i]))
+      body(cdf) = substitute(1 - (basedist$survival(x) *
+        (exp(-y) + (1 - exp(-y)) * basedist$survival(x))^-1), # nolint
+      list(y = lp[i]))
+      body(quantile) = substitute(basedist$quantile(-p / ((exp(-y) * (p - 1)) - p)), # nolint
+        list(y = lp[i]))
+      params[[i]]$pdf = pdf
+      params[[i]]$cdf = cdf
+      params[[i]]$quantile = quantile
+    }
+  }
+
+  distlist = lapply(params, function(.x) do.call(distr6::Distribution$new, .x))
+  names(distlist) = paste0(short_name, seq_along(distlist))
+
+  distr = distr6::VectorDistribution$new(distlist,
+    decorators = c("CoreStatistics", "ExoticStatistics"))
+
+  lp = lp + fit$coefficients[1]
+
+  list(lp = as.numeric(lp), distr = distr)
+}
+
+.predict_survreg_discrete <- function(object, newdata, form, predict_times,
+  basedist, fit, lp) {
+
+  if (form == "tobit") {
+    fun = function(y) stats::pnorm((predict_times - y - fit$coefficients[1]) / basedist$stdev())
+  } else if (form == "ph") {
+    fun = function(y) 1 - (basedist$survival(predict_times)^exp(-y))
+  } else if (form == "aft") {
+    fun = function(y) 1 - (basedist$survival(predict_times / exp(y)))
+  } else if (form == "po") {
+    fun = function(y) {
+      surv = basedist$survival(predict_times)
+      1 - (surv * (exp(-y) + (1 - exp(-y)) * surv)^-1)
+    }
+  }
+
+  if (length(predict_times) == 1) { # edge case
+    mat <- as.matrix(vapply(lp, fun, numeric(1)), ncol = 1)
+  } else {
+    mat <- t(vapply(lp, fun, numeric(length(predict_times))))
+  }
+  colnames(mat) <- predict_times
+
+  list(
+    lp = as.numeric(lp + fit$coefficients[1]),
+    distr = distr6::as.Distribution(mat, fun = "cdf")
```

Comment:
should here just return the survival matrix `mat`? why convert to `distr6` since its the discrete case?